Add Adafactor foreach impl

janeyx99 opened the pull request:
This PR adds the foreach impl for Adafactor knowing that there are many ways to improve its runtime perf today (by adding more foreach support). After this PR:
- we have a foreach flag for Adafactor
- It is NOT the default. Why not? It is only slightly faster + uses O(n) more memory where n is the number of params in your max param group. People tend to use Adafactor for memory efficiency.

Next steps:
- make torch.compile possible on it
- make it faster (by adding more foreach apis)

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #132336
* #133360

janeyx99 committed:
Add Adafactor foreach impl

[ghstack-poisoned]

janeyx99 committed:
Update on "Add Adafactor foreach impl"




[ghstack-poisoned]

janeyx99 committed:
Update on "Add Adafactor foreach impl"




[ghstack-poisoned]

janeyx99 committed:
Update on "Add Adafactor foreach impl"

This PR adds the foreach impl for Adafactor knowing that there are many ways to improve its runtime perf today (by adding more foreach support). After this PR:
- we have a foreach flag for Adafactor
- It is NOT the default. Why not? It is only slightly faster + uses O(n) more memory where n is the number of params in your max param group. People tend to use Adafactor for memory efficiency.

Next steps:
- make torch.compile possible on it
- make it faster (by adding more foreach apis)




[ghstack-poisoned]

janeyx99 commented on torch/optim/_adafactor.py:
```diff
@@ -395,6 +407,186 @@ def _single_tensor_adafactor(
+            ]
+            torch._foreach_mul_(row_means, row_means)
+            torch._foreach_div_(row_means, [grad.size(-1) for grad in device_grads])  # type: ignore[union-attr]
+            torch._foreach_mul_(device_row_vars, beta2_ts)  # type: ignore[arg-type]
+            torch._foreach_mul_(row_means, one_minus_beta2_ts)
+            torch._foreach_add_(device_row_vars, row_means)  # type: ignore[arg-type]
```
in the future would be a 
```
torch._foreach_lerp_(device_row_vars, row_means, one_minus_beta2_ts)
```

if we had ScalarList support for _foreach_lerp 3rd arg

albanD replied:
Do you expect to do these before merging the PR?
Should these improvements be recapped in an issue?

janeyx99 commented on torch/optim/_adafactor.py:
```diff
@@ -395,6 +407,186 @@ def _single_tensor_adafactor(
+            ]
+            torch._foreach_mul_(col_means, col_means)
+            torch._foreach_div_(col_means, [grad.size(-2) for grad in device_grads])  # type: ignore[union-attr]
+            torch._foreach_mul_(device_col_vars, beta2_ts)  # type: ignore[arg-type]
+            torch._foreach_mul_(col_means, one_minus_beta2_ts)
+            torch._foreach_add_(device_col_vars, col_means)  # type: ignore[arg-type]
```
`torch._foreach_lerp_(device_col_vars, col_means, one_minus_beta2_ts)`

janeyx99 commented on torch/optim/_adafactor.py:
```diff
@@ -395,6 +407,186 @@ def _single_tensor_adafactor(
+            ), "variance should be defined when grad is a vector"
+
+            grads_squared = torch._foreach_mul(device_grads, device_grads)  # type: ignore[arg-type]
+            torch._foreach_mul_(device_variances, beta2_ts)  # type: ignore[arg-type]
+            torch._foreach_mul_(grads_squared, one_minus_beta2_ts)
+            torch._foreach_add_(device_variances, grads_squared)  # type: ignore[arg-type]
```
`torch._foreach_lerp_(device_variances, grads_squared, one_minus_beta2_ts)`

janeyx99 commented on torch/optim/_adafactor.py:
```diff
@@ -395,6 +407,183 @@ def _single_tensor_adafactor(
+            ), "row_var and col_var should be defined when grad is multidimensional"
+            # same as (g * g).mean(dim=-1) w/o materializing an intermediate size g
+            row_means = [
+                torch.norm(grad, dim=-1, keepdim=True) for grad in device_grads
```
no foreach norm support for this type of norm

janeyx99 commented on torch/optim/_adafactor.py:
```diff
@@ -395,6 +407,183 @@ def _single_tensor_adafactor(
+                for row_var, col_var in zip(device_row_vars, device_col_vars)
+            ]
+            row_var_means = [
+                row_var.mean(dim=-2, keepdim=True) for row_var in device_row_vars  # type: ignore[union-attr]
```
no foreach mean

janeyx99 commented on torch/optim/_adafactor.py:
```diff
@@ -395,6 +407,183 @@ def _single_tensor_adafactor(
+            del col_means
+
+            var_estimates = [
+                row_var @ col_var  # type: ignore[operator]
```
no foreach mm lol, probably the bulk of the work

janeyx99 commented on torch/optim/_adafactor.py:
```diff
@@ -395,6 +407,183 @@ def _single_tensor_adafactor(
+
+        # square the eps1 as we sqrt after to keep eps1's magnitude
+        torch._foreach_clamp_min_(var_estimates, eps1 * eps1)
+        torch._foreach_sqrt_(var_estimates)
+        torch._foreach_reciprocal_(var_estimates)
```
would benefit from a foreach_rsqrt

janeyx99 commented on torch/optim/_adafactor.py:
```diff
@@ -395,6 +407,183 @@ def _single_tensor_adafactor(
+            for a, update in zip(alphas, updates)
+        ]
+        torch._foreach_mul_(updates, alphas)
+        torch._foreach_add_(device_params, updates)  # type: ignore[arg-type]
```
would be nice to have a foreach_add where the alphas could be a scalarlist

albanD commented on torch/optim/_adafactor.py:
```diff
@@ -395,6 +407,183 @@ def _single_tensor_adafactor(
+                device_state_steps, torch.tensor(1.0, device="cpu"), alpha=1.0  # type: ignore[arg-type]
+            )
+        else:
+            torch._foreach_add_(device_state_steps, 1)  # type: ignore[arg-type]
```
```suggestion
            torch._foreach_add_(device_state_steps, 1.)  # type: ignore[arg-type]
```

?

albanD reviewed (commented):
Sounds pretty good!
Curious what's the plan for all the future improvements

janeyx99 commented:
@albanD I'm planning to encapsulate all the action items in an issue before landing this PR. Including perf wins, compile support, etc.

janeyx99 committed:
Update on "Add Adafactor foreach impl"

This PR adds the foreach impl for Adafactor knowing that there are many ways to improve its runtime perf today (by adding more foreach support). After this PR:
- we have a foreach flag for Adafactor
- It is NOT the default. Why not? It is only slightly faster + uses O(n) more memory where n is the number of params in your max param group. People tend to use Adafactor for memory efficiency.

Next steps:
- make torch.compile possible on it
- make it faster (by adding more foreach apis)




[ghstack-poisoned]

janeyx99 committed:
Update on "Add Adafactor foreach impl"

This PR adds the foreach impl for Adafactor knowing that there are many ways to improve its runtime perf today (by adding more foreach support). After this PR:
- we have a foreach flag for Adafactor
- It is NOT the default. Why not? It is only slightly faster + uses O(n) more memory where n is the number of params in your max param group. People tend to use Adafactor for memory efficiency.

Next steps:
- make torch.compile possible on it
- make it faster (by adding more foreach apis)




[ghstack-poisoned]

janeyx99 commented:
Perf tracker with all issues: https://github.com/pytorch/pytorch/issues/133367

janeyx99 committed:
Update on "Add Adafactor foreach impl"

This PR adds the foreach impl for Adafactor knowing that there are many ways to improve its runtime perf today (by adding more foreach support). After this PR:
- we have a foreach flag for Adafactor
- It is NOT the default. Why not? It is only slightly faster + uses O(n) more memory where n is the number of params in your max param group. People tend to use Adafactor for memory efficiency.

Next steps:
- make torch.compile possible on it
- make it faster (by adding more foreach apis)




[ghstack-poisoned]

albanD reviewed (approved):
Sounds good!

janeyx99 commented:
@pytorchbot merge
